contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchImpl.java

```diff
@@ -0,0 +1,148 @@
+package com.linkedin.parseq.batching;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import com.linkedin.parseq.promise.PromiseResolvedException;
+import com.linkedin.parseq.promise.Promises;
+import com.linkedin.parseq.promise.SettablePromise;
+import com.linkedin.parseq.trace.ShallowTraceBuilder;
+
+public class BatchImpl<K, T> implements Batch<K, T> {
+
+  private final Map<K, BatchEntry<T>> _map;
+
+  private BatchImpl(Map<K, BatchEntry<T>> map) {
+    _map = map;
+  }
+
+  @Override
+  public void done(K key, T value) throws PromiseResolvedException {
+    _map.get(key).getPromise().done(value);
+  }
+
+  @Override
+  public void fail(K key, Throwable error) throws PromiseResolvedException {
+    _map.get(key).getPromise().fail(error);
+  }
+
+  @Override
+  public void failAll(Throwable error) throws PromiseResolvedException {
+    PromiseResolvedException exception = null;
+    for (BatchEntry<T> entry: _map.values()) {
+      try {
+        entry.getPromise().fail(error);
+      } catch (PromiseResolvedException e) {
+        exception = e;
+      }
+    }
+    if (exception != null) {
+      throw exception;
+    }
+  }
+
+  @Override
+  public Set<K> keys() {
+    return _map.keySet();
+  }
+
+  @Override
+  public int size() {
+    return _map.size();
+  }
+
+  @Override
+  public void foreach(final BiConsumer<K, SettablePromise<T>> consumer) {
+    _map.forEach((key, entry) -> consumer.accept(key, entry.getPromise()));
+  }
+
+  @Override
+  public String toString() {
+    return "BatchImpl [entries=" + _map + "]";
+  }
+
+  @Override
+  public boolean failAllRemaining(Throwable error) {
```

Comment:
API definitely needs documentation. It fails all unsatisfied promises and returns true if ann promises were failed. It returns false if there was at least one promise that has already been satisfied.
If promise is already resolved then trying to fail it will throw exception but it will not fail it.